### PR TITLE
ENT-3381: Only create new session if token expired

### PR DESCRIPTION
**Pull request by binodpant**
Why?
To avoid creating a bazillion sessions and tokens
Right now every call to create/update/delete content_metadata will issue a new session and a new auth token. While it works, it's an optimization for us to keep reusing the session and the auth token while it is still valid

This PR creates a new session when _create_session() is called (and obtains a new auth token and jams it into the session), but only if one of these is true:

1. token expiry is detected based on a stored `expires_at` attribute
2. token has never been obtained (is None)
3. session has never been initialized (is None)

If 2 and 3 are false, it still handles the case of 1, token expiry

also moved around test code for better org (and one small refactor on token response method)

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [x] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [x] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [Travis](https://travis-ci.org/github/edx/edx-enterprise), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: integrated_channels/canvas/client.py [hunk 1]
```diff
@@ -424,6 +433,6 @@ def _get_oauth_access_token(self, client_id, client_secret):
         auth_response.raise_for_status()
         try:
             data = auth_response.json()
-            return data['access_token']
+            return data['access_token'], data["expires_in"]
```
Comment (alex-sheehan-edx) [inline integrated_channels/canvas/client.py:436]: 👍  nice that it worked out this well

File: integrated_channels/canvas/client.py [hunk 2]
```diff
@@ -173,18 +174,25 @@ def _create_session(self):
...
+                self.enterprise_configuration.client_id,
+                self.enterprise_configuration.client_secret,
+            )
+            session = requests.Session()
+            session.headers['Authorization'] = 'Bearer {}'.format(oauth_access_token)
+            session.headers['content-type'] = 'application/json'
+            self.session = session
+            # expiry expected after `expires_in` seconds
+            if expires_in is not None:
+                self.expires_at = datetime.utcnow() + timedelta(seconds=expires_in)
```
Comment (alex-sheehan-edx) [inline integrated_channels/canvas/client.py:195]: I didn't look into the expiration handling from Canvas, is it always in seconds? What is the actual amount of time until we lose the token?

File: integrated_channels/canvas/client.py [hunk 3]
```diff
@@ -394,6 +402,7 @@ def _get_oauth_access_token(self, client_id, client_secret):
 
         Returns:
             access_token (str): the OAuth access token to access the Canvas API as the user
+            expires_in (int): the number of seconds after which token will expire
```
Comment (alex-sheehan-edx) [inline integrated_channels/canvas/client.py:405]: soooo nitpicky here, but `the token will expire`?

**Review by alex-sheehan-edx — APPROVED** [review]
A couple q's but otherwise this looks real good 👍

File: integrated_channels/canvas/client.py (hunk 2 above)
Comment (binodpant) [reply integrated_channels/canvas/client.py:195]: yup, always in seconds (1 hour by default)

Comment (binodpant) [reply integrated_channels/canvas/client.py:195]: https://canvas.instructure.com/doc/api/file.oauth.html#using-refresh-tokens

File: integrated_channels/canvas/client.py (hunk 3 above)
Comment (binodpant) [reply integrated_channels/canvas/client.py:405]: :)

File: integrated_channels/canvas/client.py (hunk 2 above)
Comment (binodpant) [reply integrated_channels/canvas/client.py:195]: regardless, this will use the expires_in (Seconds) as a way to setup when we should expect expiry

**Pull request merged**